mesop/server/state_session.py

```diff
@@ -269,6 +269,98 @@ def clear_stale_sessions(self):
     pass
 
 
+class SqlStateSessionBackend(StateSessionBackend):
+  """State session backend using SQL via SQL Alchemy."""
+
+  # Hardcode for now, but probably make adjustable in the future.
+  _SESSION_TTL_MINUTES = 10
+  _SESSION_CLEAR_N_REQUESTS = 100
+
+  def __init__(self, connection_uri: str, table_name: str):
+    from sqlalchemy import (
+      Column,
+      DateTime,
+      LargeBinary,
+      MetaData,
+      String,
+      Table,
+      create_engine,
+    )
+
+    self.db = create_engine(connection_uri)
+    self.metadata = MetaData()
+    self.table = Table(
+      table_name,
+      self.metadata,
+      Column("token", String(23), primary_key=True),
+      Column("states", LargeBinary, nullable=False),
+      Column("created_at", DateTime, nullable=False, index=True),
+    )
+    # Tries to create the database if it does not exist. Outside of SQLite3, the user
+    # will need to manually create the database table since the database user should
+    # generally have the minimum necessary privileges.
+    #
+    # If the table does not exist and cannot be created a sqlalchemy.exc.ProgrammingError
+    # error will be raised.
+    self.metadata.create_all(self.db)
+
+    # We will count each restore call as a request. This is used for tracking when we
+    # should clear stale database sessions.
+    self.request_count = 0
+
+  def restore(self, token: str, states: States):
+    from sqlalchemy import delete, select
+
+    self.request_count += 1
+
+    with self.db.connect() as conn:
+      result = conn.execute(
+        select(self.table).where(self.table.c.token == token)
+      ).one_or_none()
+      if result:
+        states_data = msgpack.unpackb(result.states)  # type: ignore
+        _deserialize_state(states, states_data)
+      else:
+        raise Exception("Token not found in state session backend.")
+
+    with self.db.begin() as conn:
+      conn.execute(delete(self.table).where(self.table.c.token == token))
+
+  def save(self, token: str, states: States):
+    from sqlalchemy import insert
+
+    with self.db.begin() as conn:
+      conn.execute(
+        insert(self.table).values(
+          token=token,
+          states=msgpack.packb(_serialize_state(states)),
+          created_at=_current_datetime_utc(),
+        )
+      )
+
+  def clear_stale_sessions(self):
+    """Clears stale sessions from database.
+
+    Only query the database after every X requests to avoid fruitlessly querying the
+    database unnecessarily.
+    """
+    from sqlalchemy import delete
+
+    if self.request_count < self._SESSION_CLEAR_N_REQUESTS:
```

Comment:
This doesn't seem right to me. For services like Cloud Run, it's very possible that the process only serves one request (or a few), before it's disposed.

I think we can keep this simple for now and just clear stale sessions every time. It's possible for some SQL databases, they can have a TTL policy, but I think we can wait before optimizing this.

Reply:
That's a good point. I think it would still be good to run every X requests, but just at a lower number. I've changed that to 10 for now.

I think Cloud Run instances get disposed if they are inactive for 15 minutes.

It's definitely possible that the server never hits 10 requests, but in that case, there's probably very little data there. It definitely could slowly accumulate more and more stale sessions but it would take a while to do so while never hitting 10 requests.

In future, this possible could be made configurable so the user can control based on their traffic and server.

Reply:
Good point about the 15 minute waiting period - I didn't realize it was that long. I think this is good enough for now.

I think longer term, something like [Cloud Run scheduled jobs](https://cloud.google.com/run/docs/execute/jobs-on-schedule) could be a more robust solution as it would enable cleanup on a regular cadence without contention, but it's a more complex deployment.

but I think, overall, we can wait for more feedback on our state session backends, before making more improvements.

Reply:
Nice point about Cloud Run Scheduled Jobs. I totally forgot about that feature. Another option is just using plain Cloud Scheduler (or Cron Job) and hitting an endpoint exposed by the Mesop server.

I'd agree that if I were developing this for myself, I'd set up some separate task to do the clean up rather than doing after the server request. But I  also agree about potentially more complex development. Cloud Scheduler / Cron Jobs I think are relatively easy to use so long as we could safely expose an endpoint that cleaned up the sessions (though the issue with that is that it would only work for certain types of backends (i.e. probably wouldn't work for memory and with file, we'd need to hit all the servers).

So agree about waiting for more feedback.